src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java

```diff
@@ -1082,6 +1096,126 @@ public void updateGermplasmPedigree(final Integer gid, final ProgenitorsUpdateRe
+				germplasm.getPreferredName() != null ? germplasm.getPreferredName().getNval() : StringUtils.EMPTY);
+		}).collect(Collectors.toList());
+	}
+
+	private void migrateNames(final List<Integer> gidsNonSelectedGermplasm, final Integer targetGermplasmId) {
+		final List<Integer> systemNameTypeIds = this.getSystemNameTypeIds();
+		final Germplasm germplasm = this.daoFactory.getGermplasmDao().getById(targetGermplasmId);
+		final NameDAO nameDAO = this.daoFactory.getNameDao();
+		final List<Name> names = nameDAO.getNamesByGids(gidsNonSelectedGermplasm);
+		final List<Integer> existingNameTypeIds = germplasm.getNames().stream().map(Name::getTypeId).collect(Collectors.toList());
+		for (final Name name : names) {
+			// If name is a system name (e.g. PUI, CODE1, CODE2...), do not migrate if it's already present
+			if (!(systemNameTypeIds.contains(name.getTypeId()) && existingNameTypeIds.contains(name.getTypeId()))) {
+				final Name nameToSave =
+					new Name(null, germplasm, name.getTypeId(), 0, name.getNval(), name.getLocationId(), name.getNdate(),
+						name.getReferenceId());
+				nameDAO.save(nameToSave);
+				nameDAO.makeTransient(name);
+				existingNameTypeIds.add(nameToSave.getTypeId());
+			}
+		}
+	}
+
+	private void migrateAttributes(final List<Integer> gidsNonSelectedGermplasm, final Integer targetGermplasmId,
+		final Integer variableTypeId) {
+		final AttributeDAO attributeDAO = this.daoFactory.getAttributeDAO();
+		final List<Attribute> attributesOfTargetGermplasm =
+			attributeDAO.getByGIDsAndVariableType(Arrays.asList(targetGermplasmId), variableTypeId);
+		final List<Attribute> attributesOfNonSelectedGermplasm =
+			attributeDAO.getByGIDsAndVariableType(gidsNonSelectedGermplasm, variableTypeId);
+		final List<Integer> existingAttributeTypeIds =
+			attributesOfTargetGermplasm.stream().map(Attribute::getTypeId).collect(Collectors.toList());
+		for (final Attribute attribute : attributesOfNonSelectedGermplasm) {
+			if (!existingAttributeTypeIds.contains(attribute.getTypeId())) {
+				final Attribute attributeToSave =
+					new Attribute(null, targetGermplasmId, attribute.getTypeId(), attribute.getAval(), attribute.getcValueId(),
+						attribute.getLocationId(), attribute.getReferenceId(), attribute.getAdate());
+				attributeDAO.save(attributeToSave);
+				attributeDAO.makeTransient(attribute);
+				existingAttributeTypeIds.add(attributeToSave.getTypeId());
+			}
+		}
+	}
+
+	private void migrateLots(final List<GermplasmMergeRequestDto.NonSelectedGermplasm> nonSelectedGermplasmList,
+		final Integer targetGermplasmId) {
+		final List<Integer> migrateLotsGids = nonSelectedGermplasmList.stream().filter(o -> !o.isOmit() && o.isMigrateLots())
+			.map(GermplasmMergeRequestDto.NonSelectedGermplasm::getGermplasmId).collect(
+				Collectors.toList());
+		final List<Lot> lotsToMigrate = this.daoFactory.getLotDao().getByGids(migrateLotsGids);
+		for (final Lot lot : lotsToMigrate) {
+			lot.setEntityId(targetGermplasmId);
```

Comment:
We should do batch update here since we're making the same change (ie. updating to same entity id) for all the lots to migrate